### PR TITLE
Project Management Part 3: Updating Server Requests & UI for Project Details and Editing

### PR DESCRIPTION
**Pull request by abbiehuynh**
## This PR includes:
- **Server changes:**
  - Updating GET request for project information
  - Creating PUT request to update project information
- **Client changes:**
  - Creating project details page to display individual project information 
  - Implementing project form to allow users to update project information
  - Applying toggle between displaying project info and project form through Edit button
  
 ## UI Changes
Before: only had "Edit Project" Button
After: Project details page & Project Form

**Project Details Page:**
profile-page-pr7

**Project Form:**
project-form-pr7

### Notes:
- Please review the app structure and suggest any improvements for the logic of updating the project information/ project form. 
---
Merging project-details-page into projects

### CODE REVIEW

File: server/routes/projects.js [hunk 1]
```diff
@@ -89,30 +89,197 @@ app.get('/:user_id/project/:project_id', async (req, res) => {
...
+
+            const images = imageQuery.rows.map(row => ({
+                image_id: row.image_id,
+                image_url: row.image_url,
+                image_name: row.image_name,
+                image_description: row.image_description
+            }));
+
+            return res.json({
+                ...project[0],
```
Comment (davidmedina-contact) [inline server/routes/projects.js:167]: If you only need to retrieve one record, you might want to modify your Db query to limit it to just 1

File: server/routes/projects.js [hunk 2]
```diff
@@ -89,30 +89,197 @@ app.get('/:user_id/project/:project_id', async (req, res) => {
...
+            await db.query(`DELETE FROM pattern WHERE project_id = $1;`, [projectId]);
+        }
+
+        if (otherMaterials) {
+            await db.query(`DELETE FROM other_materials WHERE project_id = $1;`, [projectId]);
+        }
+
+        if (yarns) {
+            await db.query(`DELETE FROM yarn WHERE project_id = $1;`, [projectId]);
+        }
```
Comment (davidmedina-contact) [inline server/routes/projects.js:226]: Is this necessary? This endpoint should be only about updating the project details, not other entities like patterns, otherMaterials or yarns. How are these entities related to project?

File: server/routes/projects.js [hunk 3]
```diff
@@ -89,30 +89,197 @@ app.get('/:user_id/project/:project_id', async (req, res) => {
...
+        // updates yarn table
+        if (Array.isArray(yarns) && yarns.length > 0) {
+            await Promise.all(yarns.map(yarn => {
+                return db.query(
+                    `INSERT INTO yarn(project_id, yarn_brand, yarn_color, yarn_weight, yarn_type)
+                    VALUES ($1, $2, $3, $4, $5);`,
+                    [projectId, yarn.yarn_brand, yarn.yarn_color, yarn.yarn_weight, yarn.yarn_type]
+                );
+            }));
+        }
```
Comment (davidmedina-contact) [inline server/routes/projects.js:259]: I feel like each entity/table for these should have it's own endpoint so users can update them separately from the project

File: server/routes/projects.js [hunk 4]
```diff
@@ -89,30 +89,197 @@ app.get('/:user_id/project/:project_id', async (req, res) => {
...
+                return db.query(
+                    `INSERT INTO images (project_id, image_url, image_name, image_description)
+                    VALUES ($1, $2, $3, $4)
+                    ON CONFLICT (project_id, image_name) DO UPDATE
+                    SET image_url = EXCLUDED.image_url,
+                        image_description = EXCLUDED.image_description;`,
+                    [projectId, image.image_url, image.image_name, image.image_description]
+                );
+            }));
+        }
```
Comment (davidmedina-contact) [inline server/routes/projects.js:273]: Same with images

File: server/routes/testTableRoutes.js
Comment (davidmedina-contact) [inline server/routes/testTableRoutes.js:1]: For this test suite, it assumes the DB has some data in it to be tested. A lot of devs prefer to test connections with a dummy Db that uses the same schema as the real Db and uses mock data.

https://blog.platter.dev/2021/04/28/testing-branch-databases-with-jest/

File: server/routes/projects.js (hunk 2 above)
Comment (abbiehuynh) [reply server/routes/projects.js:226]: I consider those entities to be the project details (patterns, otherMaterials, yarns, etc) as those are the things that make up the project. I decided to have them as their own tables so users would be able to insert many of those things and not just one.

File: server/routes/projects.js (hunk 3 above)
Comment (abbiehuynh) [reply server/routes/projects.js:259]: I see what you mean, I think that is something I would like to improve on after completing the mvp! Since the form is quite big it would be nice to be able to edit one part without having to open and update "everything".

File: server/routes/projects.js (hunk 1 above)
Comment (abbiehuynh) [reply server/routes/projects.js:167]: Submitted changes with commit: [c3d4580](https://github.com/abbiehuynh/my-crochet-corner/pull/7/commits/c3d45805a39052081dea1c971c4d75b841f101e9)

File: server/routes/testTableRoutes.js
Comment (abbiehuynh) [reply server/routes/testTableRoutes.js:1]: This looks really helpful to apply towards testing, I appreciate the resource! I will definitely use this as a reference when I begin testing.

Comment (abbiehuynh) [reply server/routes/testTableRoutes.js:1]: I just re-read this and realized what you meant. When I was "testing" these routes, I was just doing so to make sure everything worked before creating JOIN tables, and not actually considering this file as a test suite. But now that you mention it, I will go back and create this into a testing file and use mock data instead of the actual database to create a testing suite once I start my testing phase later this week!

File: client/src/components/home/Project.jsx [hunk 5]
```diff
@@ -1,15 +1,146 @@
...
+import { Container, ListGroup, Alert, Spinner } from 'react-bootstrap';
 import ProjectForm from './ProjectForm';
 
 const Project = () => {
+    
+    const { userId, axiosInstance } = useAuth();
+    const { projectId } = useParams();
+
+    // DELETE LATER - debugging checking userId, and project Id
+    console.log('User ID:', userId, 'Project ID:', projectId);
```
Comment (mai-repo) [inline client/src/components/home/Project.jsx:13]: This is good for error handling but overall, doesn't need to be included in the final code!

File: client/src/components/home/Project.jsx [hunk 6]
```diff
@@ -1,15 +1,146 @@
...
+        try {
+            // DELETE LATER - debugging update
+            console.log(`Updating project at: /user/${userId}/project/${projectId}`, updatedProject);
+
+            const response = await axiosInstance.put(`/user/${userId}/project/${projectId}`, updatedProject);
+
+            setProject(response.data);
+            setIsEditing(false);
+        } catch (error) {
+            console.error('Error updating project:', error);
```
Comment (mai-repo) [inline client/src/components/home/Project.jsx:55]: Make this an alert instead for users to get notice of errors.

**Comment by mai-repo** [comment]
Overall, good job with your project management! I would like to see more error handling on the front-end but that can be a secondary concern.

**Comment by abbiehuynh** [comment]
> Overall, good job with your project management! I would like to see more error handling on the front-end but that can be a secondary concern.

I will definitely note to add more error handling in the front end. I may have a PR that involves cleaning logs and improving error handling!

File: client/src/components/home/Project.jsx (hunk 6 above)
Comment (abbiehuynh) [reply client/src/components/home/Project.jsx:55]: The lines below includes error handling for the user depending on the error using useState to manage the error.
Error-Handling
